### PR TITLE
Omit protocols for native WebSockets in browsers only. Fixes TypeErrors, see issue 82

### PR DESCRIPTION
**Pull request by binarykitchen**

### CODE REVIEW

**Comment by mcollina** [comment]
Just to clarify, is this a Safari only issue? test worked fine in Chrome.
In case, would you mind adding that in the comment?

File: stream.js [hunk 1]
```diff
@@ -27,7 +31,14 @@ function WebSocketStream(target, protocols, options) {
     socket = target
   // otherwise make a new one
   } else {
-    socket = new WS(target, protocols, options)
+    // special constructor treatment for native websockets in browsers, see
+    // https://github.com/maxogden/websocket-stream/issues/82
+    if (isNative && isBrowser && !protocols) {
+      socket = new WS(target)
```
Comment (jnordberg) [inline stream.js]: With this logic I can't see any way to pass protocols without also passing options as the 3rd argument. Wouldn't that cause the same error this PR is trying to fix for anyone wanting to use specific subprotocols?

File: stream.js [hunk 2]
```diff
@@ -6,7 +6,10 @@ module.exports = WebSocketStream
 
 function WebSocketStream(target, protocols, options) {
   var stream, socket
-  var socketWrite = process.title === 'browser' ? socketWriteBrowser : socketWriteNode
+
+  var isBrowser = process.title === 'browser'
+  var isNative = global.WebSocket
```
Comment (jnordberg) [inline stream.js]: IMHO *is*Something variables should be booleans, maybe add a !! before

On a side note, why isn't the check for isBrowser enough?

Comment (binarykitchen) [inline stream.js]: ok, adding a !! noew

re: isBrowser, IMO it is not the same as isNative. it is very possible to be in a browser environment where native websocket implementations are missing (think of older browsers and iOS)

File: stream.js (hunk 1 above)
Comment (binarykitchen) [inline stream.js]: according to the rfc, there is no options parameter for native websockets, see https://www.w3.org/TR/2011/WD-websockets-20110419/#the-websocket-interface

Comment (jnordberg) [inline stream.js]: Yes, but I'm talking about the optional **protocols** argument which is in the spec.

**Comment by jnordberg** [comment]
Maybe a simpler solution to this would be to let [ws-fallback.js](https://github.com/maxogden/websocket-stream/blob/master/ws-fallback.js) drop the options argument?

**Comment by binarykitchen** [comment]
@mcollina no, it also happens on chrome under mac osx, not just safari

**Comment by binarykitchen** [comment]
@jnordberg i do not follow you - feel free to add this with or after my commit

File: stream.js (hunk 1 above)
Comment (binarykitchen) [inline stream.js]: right. have a look at the if clause `if (isNative && isBrowser && !protocols) {`

it skips when a protocol exist and executes the code in the else clause which passes on the optional protocols argument

**Comment by mcollina** [comment]
@binarykitchen I've tested with latest Chrome and latest Safari and I do not see the issue in #82.

Which version of Safari/Chrome are you having issues with?

**Comment by binarykitchen** [comment]
Chrome v31.0.1650.63 on Mac OS X 10.11.2 after running for two days without any restarts.

Haven't tested this issue on Safari.

The bug is difficult to reproduce I must admit. If you want to be 100% sure, maybe raise this on the Chromium Bug Tracker and question the native WebSocket implementation there?

File: stream.js (hunk 1 above)
Comment (deanlandolt) [inline stream.js]: Wouldn't this fail in the case where you need to pass `protocols` to a native websockets impl? From some testing in safari it looks like any call w/ `arguments.length > 2` will throw a `TypeError`.

ISTM we could just drop the `!protocols` check from the first case -- something like this:

```js
if (isNative && isBrowser) {
  socket = new WS(target, protocols)
} else {
```

WFM at least.

**Comment by deanlandolt** [comment]
@binarykitchen FWIW this issue is easy for me to repro in Safari on OSX 10.10.4 (though I've never hit it in chrome). See: https://github.com/maxogden/websocket-stream/pull/83#discussion_r50798418 -- this should let us pass `protocols` in all cases.

Another alternative is just wrapping in a `try` and falling back to the 2-arity call, w/ perhaps another `try` falling back to the 1-arity call (sans `protocols`). Kinda sloppy, but ultimately still leaves us in a predictable state.

File: stream.js (hunk 1 above)
Comment (mcollina) [inline stream.js]: I agree that we need to pass protocols from the browsers. This is used by MQTT.js (as an example).

**Comment by mcollina** [comment]
My problem here is that I cannot reproduce, and it does not seem an error that could pops up after 2 days of activity.

@deanlandolt if you can reproduce consistently, would you mind checking if passing protocols in all cases is good, i.e. if the problem is only in passing `options`?

If that's the case, I presume that is the way to go.

@mafintosh @maxogden what do you think?

**Comment by deanlandolt** [comment]
I verified (on OSX 10.10.4 in Safari 8.0.7) that passing protocols is safe when `undefined`, `null`, or `[]`. I can't say whether it would throw on specific values of protocols, but presumably we'd want it to fail loudly in these cases anyway. I also verified that passing a third option (even `undefined`) always throws.

**Comment by mcollina** [comment]
@binarykitchen would you mind updating your PR and removing the `protocols` check and passing it back to the constructor?

**Comment by deanlandolt** [comment]
I wasn't sure what the preferred workflow was for this case, so I just sent a PR on top of this commit's PR to @binarykitchen branch: https://github.com/binarykitchen/websocket-stream/pull/1

**Comment by mcollina** [comment]
you can also send another PR here, I'll get that merged and released asap.

**Comment by binarykitchen** [comment]
sorry for my late response. was away and then github was down heh.

thanks @deanlandolt for your PR. had a quick look, all looks. thanks guys!

**Pull request merged**